File: app/src/main/scala/farjs/app/FarjsData.scala
package farjs.app

import scommons.nodejs.Process.Platform

import scala.scalajs.js
import scala.scalajs.js.annotation.JSImport

trait FarjsData extends js.Object {

  def getDBFilePath(): String

  def getDataDir(): js.Array[String]
}

@js.native
@JSImport("../app/FarjsData.mjs", JSImport.Default)
object FarjsData extends js.Function1[Platform, FarjsData] {

  val instance: FarjsData = js.native

  def apply(platform: Platform): FarjsData = js.native
}

File: app/FarjsData.mjs
import os from "os";
import path from "path";
import { lazyFn } from "@farjs/filelist/utils.mjs";

const appName = "FAR.js";

/**
 * @typedef {{
 *  getDBFilePath(): string;
 *  getDataDir(): readonly string[];
 * }} FarjsData
 */

/**
 * @param {NodeJS.Platform} platform
 * @returns {readonly string[]}
 */
function getAppDataDir(platform) {
  const home = os.homedir();

  if (platform === "darwin") {
    return [home, "Library", "Application Support", appName];
  }

  if (platform === "win32") {
    const appDataDir = process.env.APPDATA;
    return appDataDir !== undefined
      ? [appDataDir, appName]
      : [home, `.${appName}`];
  }

  return [home, ".local", "share", appName];
}

/**
 * @param {NodeJS.Platform} platform
 * @returns {FarjsData}
 */
function FarjsData(platform) {
  const getDataDir = lazyFn(() => getAppDataDir(platform));

  return Object.freeze({
    getDBFilePath: lazyFn(() => path.join(...getDataDir(), "farjs.db")),

    getDataDir,
  });
}

FarjsData.instance = FarjsData(process.platform);

export default FarjsData;
